sky/shell/tracing_controller.cc

```diff
@@ -19,11 +20,85 @@ namespace shell {
 const char kBaseTraceStart[] = "{\"traceEvents\":[";
 const char kBaseTraceEnd[] = "]}";
 
+const char kObservatoryMethodStartTracing[] = "flutter_startTracing";
+const char kObservatoryMethodStopTracing[] = "flutter_stopTracing";
+const char kObservatoryResultOk[] = "{\"success\" : true}";
+const char kObservatoryResultFail[] = "{\"success\" : false}";
+
+static const char* ObservatoryInvoke(const char* method,
+                                     const char** param_keys,
+                                     const char** param_values,
+                                     intptr_t num_params,
+                                     void* user_data) {
+  if (user_data == nullptr) {
+    return strdup(kObservatoryResultFail);
+  }
+
+  auto tracingController = reinterpret_cast<TracingController*>(user_data);
+
+  if (strncmp(method, kObservatoryMethodStartTracing,
+              sizeof(kObservatoryMethodStartTracing)) == 0) {
+    tracingController->StartTracing();
+    return strdup(kObservatoryResultOk);
+  }
+
+  if (strncmp(method, kObservatoryMethodStopTracing,
+              sizeof(kObservatoryMethodStopTracing)) == 0) {
+    // Flushing the trace log requires an active message loop. However,
+    // observatory callbacks are made on a dart worker thread. We setup a
+    // message loop manually and tell the flush completion handler to terminate
+    // the loop when done
+    base::MessageLoop workerThreadLoop;
+
+    base::FilePath tempDir;
+    bool tempAccess = base::GetTempDir(&tempDir);
+    DCHECK(tempAccess) << "Must be able to access the temp directory";
+
+    base::FilePath path = tracingController->TracePathForCurrentTime(tempDir);
+
+    tracingController->StopTracing(path, true);
+
+    // Run the loop till the flush callback terminates the activation
+    workerThreadLoop.Run();
+
+    base::File file(
+        path, base::File::Flags::FLAG_OPEN | base::File::Flags::FLAG_READ);
+    int64 length = file.GetLength();
+
+    if (length == 0) {
+      return strdup(kObservatoryResultFail);
+    }
+
+    char * data = reinterpret_cast<char *>(malloc(length));
+    int lengthRead = file.Read(0, data, length);
+
+    DCHECK(length == lengthRead);
+
+    base::DeleteFile(path, false);
+
+    return data;
+  }
+
+  return strdup(kObservatoryResultFail);
+}
+
 TracingController::TracingController()
-    : view_(nullptr), picture_tracing_enabled_(false), weak_factory_(this) {
+    : picture_tracing_enabled_(false),
+      terminate_loop_on_write_(false),
+      weak_factory_(this) {
+  ManageObservatoryCallbacks(true);
 }
 
 TracingController::~TracingController() {
+  ManageObservatoryCallbacks(false);
```

Comment:
I like all registrations to be in one spot so I dont forget to remove a registration in case more handlers are added. Don't have a strong opinion about this. Will get rid of it if you do.